### DISCUSSION TITLE
None OpenSRP 2 mobile app generated resource sync via the FHIR gateway high level definition

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2579] [Ideas]
# User Stories:

1. **Syncing Practitioners**
    - The user wants to synchronize all practitioners assigned to their location (facility).
    - The user wants to synchronize all practitioners assigned to the child locations (zones) of their assigned location.
2. **Syncing Locations**
    - The user wants to synchronize all locations (facilities).
    - The user wants to synchronize all child locations (zones) of their assigned location.

# Data Linkage

1. **Syncing Practitioners**
    - For each Practitioner resource, a corresponding Group resource is created with the `Practitioner.id` referenced in the `Group.member` attribute.
      - This Group resource is used to link the Practitioner to a CarePlan resource in cases where the Practitioner is the CarePlan subject.
    - When creating a Practitioner resource, a PractitionerRole resource is also created.
      - This resource is used to link the Practitioner to an Organization resource in cases where the Practitioner is the Organization (Team) member.
      - The PractitionerRole resource also highlights the actual role of the Practitioner in the Organization (Team) using the Practitioner or Supervisor role.
    - The Practitioner is then assigned to the CareTeam by adding the Practitioner reference to the `CareTeam.participant.member` attribute.
      - The CareTeam is assigned to the Organization (Team) by adding an Organization (Team) reference to the `CareTeam.managingOrganization` attribute.
      - The Organization (Team) reference is also added to the `CareTeam.participant.member` attribute of the CareTeam resource for easy search.
    - The Organization (Team) is assigned to a location via the OrganizationAffiliation resource.
      - The Organization is referenced on the `OrganizationAffiliation.organization` attribute.
      - The location is referenced on the `OrganizationAffiliation.location` attribute.
    - The Location child <> parent relationship is defined by the `Location.partOf` attribute.
      - The parent location is referenced on the child's `Location.partOf` attribute.
2. **Syncing Locations** (*Required for micro planning applications. Requirements not complete*)
    - The Location child <> parent relationship is defined by the `Location.partOf` attribute.
      - The parent location is referenced on the child's `Location.partOf` attribute.

# Data Syncing

1. **Syncing Practitioners**
    - Practitioners are the main clients here. Therefore, supervisors need to sync down all Practitioners in the following ways:
        - Sync all practitioners assigned to their location (facility) and all practitioners assigned to the child locations (zones) of their assigned location.
        - Sync all practitioners in the application (rare occasion).
    - Supervisors can be identified by checking the PractitionerRole resource. The PractitionerRole resource will have the Supervisor `code`, i.e., `{"coding":[{"system":"<http://snomed.info/sct","code":"236321002","display":"Supervisor> (occupation)"}]}`.
    - A config is needed to define which practitioner details to sync down. This allows the config authors to only query the data they require to build out their views on the application.
    - How to identify the Practitioners to sync down:
        1. Get the logged-in user's location (facility).
        2. Get all child locations (zones) of the logged-in user's location (facility).
        3. Using the Data linkage defined above, determine all the practitioners assigned to child locations.
    - How to identify the Practitioner details to sync down:
        1. Based on the config, get the Practitioner details to sync down.
        2. Query the HAPI server to return each Resource type based on Data linkage defined above.
2. **Syncing Locations**
    - Locations are the main clients here. Therefore, supervisors need to sync down all Locations in the following ways:
        - Sync all assigned locations (facilities).
            - Sync all child locations (zones) of their assigned location.
        - Sync all locations in the application (rare occasion).
    - A config is needed to define which Location details to sync down. This allows the config authors to only query the data they require to build out their views on the application.
    - How to identify the Locations to sync down:
        1. Get the logged-in user's location (facility).
        2. Get all child locations (zones) of the logged-in user's location (facility).

# Assumptions

1. The Practitioner and Location are the main none OpenSRP 2 mobile application created resources to sync. 
    - This assumption is based on the applications we've built on OpenSRP 1.
2. The Location and Practitioner resources are created with the above-defined Data Linkage.

**Comment by pld** [comment]
Note this may resolve the loading arbitrary data onto device part of this story, https://www.notion.so/onaio/Support-dynamically-load-data-e-g-Locations-into-Questionnaire-answer-options-2d18724fccca4345a868c4732893fb37?pvs=4

**Comment by ndegwamartin** [comment]
@dubdabasoduba had some clarifying questions:
1. For both options of syncing only attributed locations and only attributed practitioners, when is this configured? is it the same for the user through out their installation or this can change dynamically from backend configuration and at any time?

2. For the case on loading **all locations** on the device - do we have the flexibility of exploring options that side load the locations e.g. via SD card or if distribution is by _Playstore_ using [APK Expansion files](https://developer.android.com/google/play/expansion-files)?

**Reply by dubdabasoduba** [reply]
@ndegwamartin 
- The MVP implementation plan was updated to only cover the sync of practitioners.  The idea is to have a way to sync all the practitioner details for practitioners under a supervisor.

**Reply by ndegwamartin** [reply]
Noted. In that context could you respond on point 1 above?

**Comment by dubdabasoduba** [comment]
Updating this to document the technical approach
- Add the supervisor ROLE on KeyCloak.  The role will be named _SUPERVISOR_
- Add a new endpoint similar to [this](https://github.com/opensrp/fhircore/blob/ad362026de2facf8e1ba25af32d49ffca981ce97/android/engine/src/main/java/org/smartregister/fhircore/engine/util/extension/StringExtensions.kt#L106). The only difference is the request param. This I will be `location=<supervisor-uuid>`
    - This endpoint will only be triggered if the logged-in user has the _SUPERVISOR_ role.  
- Update practitioner details on the OpenSRP HAPI FHIR Extensions to add the ability to fetch multiple practitioners based on the supervisor's location.  
    - The endpoint should return all the following resources. 
        -  Group
        - PractitionerRole
        - Practitioner
        - Organization
        - CareTeam
        - Location
        - OrganizationAffiliation


cc: @ndegwamartin

**Reply by dubdabasoduba** [reply]
- This approach will only work for loading practitioner data into the app it does work for all of the other management resources. 
- This will only work for apps that are using the `Location` sync strategy
